### ISSUE TITLE
Pass initial state to useReducer

### ISSUE THREAD

**Issue by Wildhoney** [dai-shi/react-tracked#11]
First and foremost, awesome idea with using `Proxy` to track conditional component updates.

How would I go about passing a dynamic initial state into `useReducer`? Something like the following:

```javascript
const useValue = value => useReducer(store.reducer, value);

export function StoreProvider({ value, children }) {
    return <Provider useValue={() => useValue(value)}>{children}</Provider>;
}
```

In the above case I get:

> useValue must be statically defined

**Comment by dai-shi** [comment]
Hi, can you simply ignore if the `value` changes after the initialization?
I guess it's not idiomatic, but technically you could do this.

```jsx
export function StoreProvider({ value, children }) {
  const useValue = useMemo(() => useReducer(store.reducer, value), []);
  return <Provider useValue={useValue}>{children}</Provider>;
}
```

This would break the exhaustive-deps rule.

Hm, I wonder if there's a better way.

```jsx
const initialValue = ...;
const useValue = () => useReducer((state, action) => {
  if (action.type === 'INITIALIZE') return action.value;
  return store.reducer(state, action);
}, initialValue);

const InitializeValue = ({ value, children }) => {
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch({ type: 'INITIAZLIE', value });
  }, [value, dispatch]);
  return children;
};

export function StoreProvider({ value, children }) {
  return (
    <Provider useValue={useValue}>
      <InitializeValue value={value}>
        {children}
      </InitializeValue>
    </Provider>
  );
}
```

Not very beautiful though.

**Comment by Wildhoney** [comment]
Yep I did begin to spec out the second approach, but agreed it's not the most beautiful. Also the store would be initialising async so tests would need to be async too, and may also introduce FOUC.

**Comment by dai-shi** [comment]
I'd like to ask your use case, because I'm thinking of dropping the defaultContext APIs and only provide `createContainer` which might break your use case. I might re-consider.

**Comment by dai-shi** [comment]
> store would be initialising async

Oh...

**Comment by Wildhoney** [comment]
Well, in our case we're using NextJS and have the data from the `async getInitialProps` method upon load (sync from SSR) such as a user's preferences. We drop that data into the `Provider` as in the app you can update the preferences which would affect a multitude of components.

**Comment by dai-shi** [comment]
Does that mean even if you use React bare createContext/useContext, you still have the same issue?

**Comment by Wildhoney** [comment]
No it's perfectly fine, just unrelated updates to the `Provider` cause unnecessary renders on components that don't need that update.

If I could continue to pass a non-statically defined `initialState` to `useReducer` then everything would be perfect.

**Comment by dai-shi** [comment]
So, the first approach in my first comment solves the issue technically, correct?
Only the question is if it's acceptable.

**Comment by Wildhoney** [comment]
```javascript
const useValue = useMemo(() => useReducer(store.reducer, value), []);
return <Provider useValue={useValue}>{children}</Provider>;
```

Gives:

> TypeError: useValue is not a function

The only way it does *kind* of work (albeit not in tests) is with the peculiar:

```javascript
const storeInitialiser = useReducer(store.reducer, value);
const [useValue] = useState(() => () => storeInitialiser);
return <Provider useValue={useValue}>{children}</Provider>;
```

**Comment by Wildhoney** [comment]
Okay the following approach does seem to work in tests as well:

```javascript
const storeInitialiser = useReducer(store.reducer, value);
const useValue = useMemo(() => () => storeInitialiser, []);
return <Provider useValue={useValue}>{children}</Provider>;
```

Does look kind of quirky, however.

**Comment by dai-shi** [comment]
> TypeError: useValue is not a function

My bad. It should be:

```jsx
const useValue = useMemo(() => () => useReducer(store.reducer, value), []);
return <Provider useValue={useValue}>{children}</Provider>;
```

**Comment by Wildhoney** [comment]
Thanks @dai-shi. All seems good with that approach.

**Comment by Wildhoney** [comment]
Perhaps we should add it as an example to the repo? I'm sure I won't be the only one perplexed on how to achieve that.

**Comment by dai-shi** [comment]
I'm thinking of dropping `<Provider useValue={...}>` because this doesn't seem to be recommended from the react team.
To mitigate the problem, I added the "useValue must be statically defined" warning.
But, you proved it's not that helpful, so the current API should be re-considered.

As I learned from your use case, I would provide a `useReducer`-specific provider instead.
```jsx
<Provider reducer={...} initialArg={...} init={...}>
```

This would not allow non-useReducer patterns, but those can hopefully handled either by a) converting to reducer, or b) using `createContainer`.

How does that sound?
